File: Reverse_Linked_List_II.cpp
/**
 * Definition of singly-linked-list:
 * 
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *        this->val = val;
 *        this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The head of linked list.
     * @param m: The start position need to reverse.
     * @param n: The end position need to reverse.
     * @return: The new head of partial reversed linked list.
     */
    ListNode *reverseBetween(ListNode *head, int m, int n) {
        // write your code here
        ListNode *dummy = new ListNode(INT_MIN);
        dummy->next = head;
        head = dummy;
        //move m-1 steps
        for (int i = m - 1; i > 0; --i){
            head = head->next;
        }
        auto tail = head->next;
        head->next = nullptr;//cut down
        //reverse n - m + 1 nodes
        auto pre = tail;
        for (int i = n - m + 1; i > 0; --i){
            auto temp = head->next;
            head->next = pre;
            pre = pre->next;
            head->next->next = temp;
            
        }
        //linked the tail with left
        tail->next = pre;
        return dummy->next;
    }
};

File: Reverse Linked List II.py
"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The head of linked list
    @param m: start position
    @param n: end position
    """
    def reverseBetween(self, head, m, n):
        # write your code here
        
        dummy = ListNode(0)
        left = dummy
        dummy.next = head
        for i in range(m - 1):
            left = left.next
        right = left.next
        end = right
        for i in range(n - m + 1):
            # error:dont use this
            # right.next, left.next, right = left.next, right, right.next
            tmp = right
            right = right.next
            tmp.next = left.next
            left.next = tmp
        end.next = right
        return dummy.next
